qualtran/qir_interop/bloq_to_qir.py

```diff
@@ -0,0 +1,392 @@
+        irs = soq_map[soquet]
+        return [ir_map[ir] for ir in irs]
+    irs = irs_from_soquet(soquet)
+    return [ir_map[ir] for ir in irs]
+
+
+def irs_from_soquet(soq: Soquet) -> List[Tuple[str, int]]:
+    """
+    Get the IRs from a soquet.
+
+    Args:
+        soq: The soquet to get the IRs from.
+
+    Returns:
+        List of IRs from the soquet.
+    """
+    reg_name = soq.reg.name
+    starting_index = soq.idx[0] if len(soq.idx) != 0 else 0
+    return [(reg_name, starting_index * soq.reg.bitsize + i) for i in range(soq.reg.bitsize)]
+
+
+def compile_bloq(
+    bloq: qualtran.Bloq,
+    qubit_type: Type,
+    void_type: Type,
+    module: pyqir.Module,
+    context: pyqir.Context,
+    builder: Builder,
+    qubit_alloc: Function,
+    func_dict: dict,
+) -> Tuple[Function, dict]:
+    """
+    Compile a Bloq into a QIR function.
+
+    This function, at a high level, takes in a Bloq, creates a corresponding QIR function,
+    and then recursively goes through its sub-bloqs until it reaches a base case in which the bloq
+    can't be decomposed any further. The intent is for these leaf bloqs to be in the PYQIR_OP_MAP (defined above)
+    in which case the corresponding QIR operation is called. In the case where the leaf bloq is not in PYQIR_OP_MAP,
+    the leaf bloq is mapped to a QIR function without an implementation.
+
+    Args:
+        bloq: Bloq to compile.
+        qubit_type: Qubit type for the QIR function.
+        void_type: Void type for the QIR function.
+        module: Module to add the QIR function to.
+        context: Context to use.
+        builder: Builder to use.
+        qubit_alloc: Function to allocate qubits.
+        func_dict: Dictionary to store the functions already created.
+
+    Returns:
+        Tuple containing the QIR function for the given Bloq and the IR map for the Bloq.
+    """
+    func_name = f"{bloq.pretty_name()}_{get_num_qubits_for_bloq(bloq)}"
+    if func_name in func_dict:
+        return func_dict[func_name]
+    bloq_func = create_func_for_bloq(bloq, func_name, qubit_type, void_type, module)
+    ir_map = create_ir_map(bloq)
+    soq_map = {}
+
+    if not bloq_decomposes(bloq):
+        return bloq_func, ir_map
+
+    # Create a block to insert instructions in
+    basic_block = BasicBlock(context, "block1", bloq_func)
+
+    # iterate through the DAG of sub bloqs
+    for sub_bloq in bloq.decompose_bloq().iter_bloqsoqs():
+        bloq_instance, inputs, outputs = sub_bloq
+        if bloq_instance.bloq.short_name() == 'Split':
+            reg_name = inputs['reg'].reg.name
+            for i in range(len(outputs[0])):
+                ir = soq_map[inputs['reg']][i] if inputs['reg'] in soq_map else (reg_name, i)
+                soq_map[outputs[0][i]] = [ir]
+            continue
+        if bloq_instance.bloq.short_name() == 'Join':
+            irs = [el for soq in inputs['reg'] for el in soq_map[soq]]
+            soq_map[outputs[0]] = irs
+            continue
```

Comment:
We have dealt with a lot of this logic of maintaining a mapping between soquets and a flat list of qubits in the Qualtran -> Cirq conversion code. I think it may be worth seeing if we can leverage some of that here directly. 

My instinct would be to define a custom `QIRQubitManager` class that derives from `QubitManager` and keeps track of all the `qalloc` / `qfree` / `qborrow` calls and delegates to corresponding PyQIR function calls like `builder.call(qubit_alloc, [])`. 

Then, given a composite bloq `cbloq`, we can call `cbloq.to_cirq_circuit(qubit_manager=qir_qubit_manager, **input_qubits)` to generate a cirq circuit with all Bloqs. This takes care of iterative through the `cbloq`, mapping all multi-dimensional soquets etc. correctly so that we have an expanded cirq style `Operation` for each `Bloq` with a simple list of qubits that can be obtained via `op.qubits` that the bloq operators upon. Mapping this to a list of QIR function calls should then be fairly straightforward and we can hopefully get rid of `soq_map` and all the logic to maintain / update it. 

Happy to leave it as a follow-up improvement for this PR, but we should open an issue to track it.